pkg/agent/lease_counter.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package agent
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/clock"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/agent/metrics"
+	"sigs.k8s.io/apiserver-network-proxy/pkg/util"
+
+	coordinationv1api "k8s.io/api/coordination/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coordinationv1lister "k8s.io/client-go/listers/coordination/v1"
+)
+
+// A ServerLeaseCounter counts leases in the k8s apiserver to determine the
+// current proxy server count.
+type ServerLeaseCounter struct {
+	leaseLister   coordinationv1lister.LeaseLister
+	selector      labels.Selector
+	fallbackCount int
+	pc            clock.PassiveClock
+}
+
+// NewServerLeaseCounter creates a server counter that counts valid leases that match the label
+// selector and provides the fallback count (initially 0) if this fails.
+func NewServerLeaseCounter(pc clock.PassiveClock, leaseLister coordinationv1lister.LeaseLister, labelSelector labels.Selector) *ServerLeaseCounter {
+	return &ServerLeaseCounter{
+		leaseLister:   leaseLister,
+		selector:      labelSelector,
+		fallbackCount: 1,
+		pc:            pc,
+	}
+}
+
+// Count counts the number of leases in the apiserver matching the provided
+// label selector.
+//
+// In the event that no valid leases are found or lease listing fails, the
+// fallback count is returned. This fallback count is updated upon successful
+// discovery of valid leases.
+func (lc *ServerLeaseCounter) Count() int {
+	// Since the number of proxy servers is generally small (1-10), we opted against
+	// using a LIST and WATCH pattern and instead list all leases on each call.
+	// TODO: Switch to an informer-based system and use events rather than a polling loop.
```

Comment:
don't need the metric writes here anymore

Reply:
Removed.